### PR TITLE
Spearbit 23

### PR DESCRIPTION
**Pull request by liu-zhipeng**
## Type of change

- [ ] Docs change / dependency upgrade
- [ ] Configuration / tooling changes
- [x] Refactoring
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Requires changes in customer code

## Related Issue(s)

Fixes 
https://github.com/spearbit-audits/connext/issues/23

### CODE REVIEW

**Review by LayneHaber — APPROVED** [review]

**Review by LayneHaber — CHANGES_REQUESTED** [review]
See: https://github.com/spearbit-audits/connext/issues/23#issuecomment-1196519233

File: packages/deployments/contracts/contracts/core/connext/libraries/SwapUtils.sol
Comment (gpersoon) [inline packages/deployments/contracts/contracts/core/connext/libraries/SwapUtils.sol:211]: I think the last past should be:
`( xpi - ((xpi * v.d1) / v.d0))) * (v.feePerToken)) /`

File: packages/deployments/contracts/contracts/core/connext/libraries/SwapUtils.sol
Comment (0xleastwood) [inline packages/deployments/contracts/contracts/core/connext/libraries/SwapUtils.sol:460]: I think `y = ((y * y) + c) / ((y * 2) + b - d);` would be safer

**Review by LayneHaber — APPROVED** [review]
LGTM pending final approval from spear bit

File: packages/deployments/contracts/contracts/core/connext/libraries/SwapUtils.sol
Comment (liu-zhipeng) [reply packages/deployments/contracts/contracts/core/connext/libraries/SwapUtils.sol]: hmm. looks good

File: packages/deployments/contracts/contracts/core/connext/libraries/SwapUtilsExternal.sol
Comment (0xbok) [inline packages/deployments/contracts/contracts/core/connext/libraries/SwapUtilsExternal.sol:872]: this line can be moved back here?

File: packages/deployments/contracts/contracts/core/connext/libraries/SwapUtilsExternal.sol
Comment (0xbok) [inline packages/deployments/contracts/contracts/core/connext/libraries/SwapUtilsExternal.sol:508]: same here: (y * y) and (y * 2).

File: packages/deployments/contracts/contracts/core/connext/libraries/SwapUtils.sol
Comment (0xbok) [inline packages/deployments/contracts/contracts/core/connext/libraries/SwapUtils.sol:276]: same here: (y * y) and (y * 2).

**Comment by 0xbok** [comment]
LGTM, some minor comments. In general, parenthesis can be used to have explicit order of operations, but I didn't find any discrepancies in the arithmetic before and after. You can refer to the order of operations here: https://docs.soliditylang.org/en/v0.8.15/cheatsheet.html?highlight=order%20of%20operation#order-of-precedence-of-operators

File: packages/deployments/contracts/contracts/core/connext/libraries/SwapUtils.sol
Comment (liu-zhipeng) [reply packages/deployments/contracts/contracts/core/connext/libraries/SwapUtils.sol:223]: what is your suggestion? @0xbok

Comment (0xbok) [reply packages/deployments/contracts/contracts/core/connext/libraries/SwapUtils.sol:223]: it might be a personal preference, but I think it'll be easier to parse by eyes if `* v.feePerToken` is in a different line. But feel free to ignore this comment entirely.

File: packages/deployments/contracts/contracts/core/connext/libraries/AmplificationUtils.sol
Comment (0xleastwood) [inline packages/deployments/contracts/contracts/core/connext/libraries/AmplificationUtils.sol:61]: Based on https://docs.soliditylang.org/en/v0.8.15/cheatsheet.html?#order-of-precedence-of-operators isn't `((a0 - a1) * (block.timestamp - t0)) / (t1 - t0);` evaluated first. Should this be `(a0 + ((a0 - a1) * (block.timestamp - t0))) / (t1 - t0);`?

Comment (0xleastwood) [reply packages/deployments/contracts/contracts/core/connext/libraries/AmplificationUtils.sol:61]: Just want to confirm that its intended that `a0` is added to the result of `((a0 - a1) * (block.timestamp - t0)) / (t1 - t0);` and not added to the numerator before division.

Comment (liu-zhipeng) [reply packages/deployments/contracts/contracts/core/connext/libraries/AmplificationUtils.sol:61]: https://github.com/curvefi/curve-contract/blob/b0bbf77f8f93c9c5f4e415bce9cd71f0cdee960e/contracts/pools/3pool/StableSwap3Pool.vy#L160

I believe that is correct.

Comment (0xleastwood) [reply packages/deployments/contracts/contracts/core/connext/libraries/AmplificationUtils.sol:61]: cool, that looks consistent then.

**Pull request merged**
